src/db/resolver.rs

```diff
@@ -191,3 +192,211 @@ RETURNING id
 
     Ok(deleted_ids.try_into().unwrap())
 }
+
+/// Function to delete messages older than `retention` minutes
+/// Returns the number of messages deleted
+pub async fn prune_old_messages(pool: &PgPool, retention: i32) -> Result<i64, anyhow::Error> {
+    let cutoff_nonce = Utc::now().timestamp() - (retention as i64 * 60);
+
+    let deleted_count = sqlx::query!(
+        r#"
+        DELETE FROM messages
+        WHERE (message->>'nonce')::bigint < $1
+        RETURNING id
+        "#,
+        cutoff_nonce
+    )
+    .fetch_all(pool)
+    .await?
+    .len() as i64;
+
+    Ok(deleted_count)
+}
+
+pub async fn list_active_indexers(
+    pool: &PgPool,
+    indexers: Option<Vec<String>>,
+    from_timestamp: i64,
+) -> Result<Vec<String>, anyhow::Error> {
+    let mut query = String::from("SELECT DISTINCT message->>'graph_account' as graph_account FROM messages WHERE (CAST(message->>'nonce' AS BIGINT)) > $1");
+
+    // Dynamically add placeholders for indexers if provided.
+    if let Some(ref idxs) = indexers {
+        let placeholders = idxs
+            .iter()
+            .enumerate()
+            .map(|(i, _)| format!("${}", i + 2))
+            .collect::<Vec<_>>()
+            .join(",");
+        query.push_str(&format!(
+            " AND (message->>'graph_account') IN ({})",
+            placeholders
+        ));
+    }
+
+    let mut query = sqlx::query(&query).bind(from_timestamp);
+
+    // Bind indexers to the query if provided.
+    if let Some(indexers) = indexers {
+        for account in indexers {
+            query = query.bind(account);
+        }
+    }
+
+    let rows = query
+        .fetch_all(pool)
+        .await
+        .map_err(anyhow::Error::new)? // Convert sqlx::Error to anyhow::Error for uniform error handling.
+        .iter()
+        .map(|row| row.get::<String, _>("graph_account"))
+        .collect();
+
+    Ok(rows)
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use chrono::Utc;
+    use serde_json::json;
+    use sqlx::{PgPool, migrate::Migrator};
+    use std::env;
+
+    static MIGRATOR: Migrator = sqlx::migrate!();
+
+    async fn setup_test_db() -> PgPool {
```

Comment:
could you explain why is this fn used instead of `sqlx::test`?

Reply:
✅